tools/analyzer_plugin/lib/src/util/prop_declarations/required_props.dart

```diff
@@ -0,0 +1,122 @@
+    // (either an explicit override or not) that we processed in a previous iteration.
+    // Since it could have a different requiredness, use the maximum requiredness between the two.
+    final requiredness = maxRequiredness(_getPropRequiredness(propField), propRequirednessByName[name]);
+
+    propRequirednessByName[name] = requiredness;
+    if (requiredness.isRequired) {
+      requiredFieldsByName[name] = propField;
+    }
+  }
+  return RequiredPropInfo._(requiredFieldsByName, propRequirednessByName);
+}
+
+/// A collection of all props declared in a given props class and its supertypes,
+/// and whether each prop is required.
+///
+/// Returned by [getAllRequiredProps].
+class RequiredPropInfo {
+  /// A mapping of prop field names to the fields that declare them, containing only required props.
+  final Map<String, FieldElement> requiredFieldsByName;
+
+  /// A mapping of all prop field names to their requiredness.
+  final Map<String, PropRequiredness> propRequirednessByName;
+
+  RequiredPropInfo._(this.requiredFieldsByName, this.propRequirednessByName);
+
+  /// The prop field names of all required props.
+  Iterable<String> get requiredPropNames => requiredFieldsByName.keys;
+}
+
+/// Returns the requiredness for a given prop field declared by [propField].
+PropRequiredness _getPropRequiredness(FieldElement propField) {
+  if (propField.isLate) {
+    return PropRequiredness.late;
+  }
+
+  if (propField.metadata.any((annotation) {
+    if (annotation.element?.library?.name != 'over_react.component_declaration.annotations') {
+      return false;
+    }
+    // It's almost always going to be `requiredProp` or `Accessor`;
+    // skip more checks and just try to pull the bool field `isRequired` off of the constant value.
+    // If it fails, it's not the annotation we're interested in anyways.
+    return annotation.computeConstantValue()?.getField('isRequired')?.toBoolValue() ?? false;
+  })) {
+    return PropRequiredness.annotation;
+  }
+
+  return PropRequiredness.none;
+}
+
+/// A value representing whether a prop is required, and how it's required.
+enum PropRequiredness {
+  /// Represents a prop required via late keyword (only available in null-safe language versions).
+  ///
+  /// When missing, throws upon UiProps invocation when asserts are enabled when invoking a builder (new boilerplate only).
+  late(isRequired: true, requirednessLevel: 3),
+
+  /// Represents a prop required via annotations: `@requiredProp`, `@requiredNullableProp`, or `@Accessor(isRequired: true)`.
+  ///
+  /// When missing, will either throw, warn, or do nothing at component render, depending on the component boilerplate.
+  ///
+  /// Deprecated in favor of [late].
+  annotation(isRequired: true, requirednessLevel: 2),
+
+  /// Represents a prop that was considered required in its declaration,
+  /// but is ignored by the consuming class.
+  ignoredByConsumingClass(isRequired: false, requirednessLevel: 1),
```

Comment:
This is currently unused, but will get used as part of updates to this file in FED-2015; since I was adding `requirednessLevel` and new `PropRequiredness` enum test cases, I figured I'd just stub it in now.